pkg/gds/gds_test.go

```diff
@@ -322,55 +322,86 @@ func (s *gdsTestSuite) TestVerifyContact() {
 	defer s.grpc.Close()
 	client := api.NewTRISADirectoryClient(s.grpc.Conn)
 
-	// VASP does not exist in the database
+	charlieID := s.fixtures[vasps]["charliebank"].(*pb.VASP).Id
+
+	// Cannot verify contact without a token
 	request := &api.VerifyContactRequest{
-		Id:    "abc12345-41aa-11ec-9d29-acde48001122",
-		Token: "",
+		Id: charlieID,
 	}
 	_, err := client.VerifyContact(ctx, request)
 	require.Error(err)
 
+	// VASP does not exist in the database
+	request = &api.VerifyContactRequest{
+		Id:    "abc12345-41aa-11ec-9d29-acde48001122",
+		Token: "administrative_token",
+	}
+	_, err = client.VerifyContact(ctx, request)
+	require.Error(err)
+
 	// Incorrect token - no verified contacts
-	request.Id = s.fixtures[vasps]["charliebank"].(*pb.VASP).Id
+	request.Id = charlieID
 	request.Token = "invalid"
 	_, err = client.VerifyContact(ctx, request)
 	require.Error(err)
 
-	// TODO: Test previously verified contact - requires modifying the fixtures to
-	// include a non-empty verification token
-
 	// Successful verification
-	request.Token = ""
+	request.Token = "administrative_token"
 	sent := time.Now()
 	reply, err := client.VerifyContact(ctx, request)
 	require.NoError(err)
 	require.Nil(reply.Error)
 	require.Equal(pb.VerificationState_PENDING_REVIEW, reply.Status)
 	require.Contains(reply.Message, "successfully verified")
-
 	// VASP on the database should be updated
 	vasp, err := s.svc.GetStore().RetrieveVASP(request.Id)
 	require.NoError(err)
 	require.Equal(pb.VerificationState_PENDING_REVIEW, vasp.VerificationStatus)
 	token, err := models.GetAdminVerificationToken(vasp)
 	require.NoError(err)
 	require.NotEmpty(token)
+	token, verified, err := models.GetContactVerification(vasp.Contacts.Administrative)
+	require.NoError(err)
+	require.Empty(token)
+	require.True(verified)
+
+	// Verify a different contact
+	request.Token = "legal_token"
+	reply, err = client.VerifyContact(ctx, request)
+	require.NoError(err)
+	require.Nil(reply.Error)
+	require.Equal(pb.VerificationState_PENDING_REVIEW, reply.Status)
+	// Should only change the fields on the contact
+	vasp, err = s.svc.GetStore().RetrieveVASP(request.Id)
+	require.NoError(err)
+	require.Equal(pb.VerificationState_PENDING_REVIEW, vasp.VerificationStatus)
+	token, verified, err = models.GetContactVerification(vasp.Contacts.Legal)
+	require.NoError(err)
+	require.Empty(token)
+	require.True(verified)
+
+	// Attempt to verify an already verified contact - should fail
+	request.Token = "legal_token"
+	_, err = client.VerifyContact(ctx, request)
+	require.Error(err)
 
-	// Audit log should contain new entries for contact verifications, EMAIL_VERIFIED,
-	// PENDING_REVIEW, along with the intitial SUBMITTED.
+	// Check audit log entries
 	log, err := models.GetAuditLog(vasp)
 	require.NoError(err)
-	// Currently verifies all contacts because the fixtures all have the empty token.
-	require.Len(log, 7)
+	require.Len(log, 5)
```

Comment:
Just confirming that this dropped from 7 to 5 because only 5 have verified contacts?

Reply:
With the previous fixtures we were basically abusing an edge case in `VerifyContact` where an empty token string was treated as valid, meaning all four of the contacts were verified. With this PR, only two of the contacts are verified. The audit log has 5 entries in total because of the initial `SUBMITTED` state in the fixture, 2 for contact verifications (although the VASP state is unchanged), and 2 for changing the VASP state to `EMAIL_VERIFIED` and `PENDING` during the contact verification process.